project/Build.scala

```diff
@@ -1253,13 +1318,40 @@ object Build {
 
       cleanIRSettings,
 
+      Compile / doc := {
+        val dir = (Compile / doc / target).value
+        IO.createDirectory(dir)
+        dir
+      },
+
       headerSources in Compile ~= { srcs =>
         srcs.filter { src =>
           val path = src.getPath.replace('\\', '/')
           !path.contains("/java/math/") &&
           !path.endsWith("/java/util/concurrent/ThreadLocalRandom.scala")
         }
       },
+  ).withScalaJSCompiler2_12.dependsOnLibraryNoJar2_12
+
+  /** An empty project, without source nor dependencies, whose products are
+   *  copied from `javalibInternal`.
+   *
+   *  This the "public" version of the javalib, as depended on by the `library`
+   *  and published on Maven.
+   */
+  lazy val javalib: Project = Project(
+      id = "javalib", base = file("javalib-public")
+  ).settings(
+      commonSettings,
+      name := "scalajs-javalib",
+      publishSettings(Some(VersionScheme.BreakOnMajor)),
+
+      crossPaths := false,
+      autoScalaLibrary := false,
+      crossVersion := CrossVersion.disabled,
+
+      Compile / packageBin / mappings := (javalibInternal / Compile / packageBin / mappings).value,
+      exportJars := false, // very important, otherwise there's a cycle with the `library`
```

Comment:
I'm wondering whether this wouldn't be made simpler if we split the `library` into two projects (and not the javalib). Nothing depends on the javalib only at compile time, so the distinction there is only to break cycles. But for the library, the distinction is real.

I could imagine something like this could work:

```scala
val libraryInternal = MultiScalaProject() // no dependencies
val scalalib = MultiScalaProject().dependsOn(libraryInternal % "provided")
val libraryAux = MultiScalaProject().dependsOn(libraryInternal % "provided")
val javalib = project.dependsOn(libraryInternal.v2_12 % "provided")

val library = MultiScalaProject().dependsOn(javalib).settings(
  mappings in packageBin := ...
)
```

Maybe I'm missing something?

Reply:
I believe that could work as well, but I think it will be more trouble than the current split. I may be wrong.

From what I understood of my experiments, even with `libraryInternal.v2_12 % Provided`, there would remain traces of the dependency in the POM file of the `javalib`. To be fully clear of the dependency, we'd have to postprocess the POM file of the javalib in that case (but not necessarily the Ivy, which is the most trouble, since that remains only in `publishLocal`).

It would relieve the burden of the artificial dependency on the `javalib` in `dependsOnLibrary`, but only to be compensated by something similar: for the projects depending on the `library`, we'll want them to declare a dependency on `library` for building and publishing, but on the `libraryInternal` for the IDE imports.

Should I give it a shot anyway?

Reply:
> Should I give it a shot anyway?

Nah, I think we should go with this in this case.